Frameworks/UIKit/NSTextContainer.mm

```diff
@@ -62,38 +78,121 @@ - (NSArray*)exclusionPaths {
     return _exclusionPaths;
 }
 
+// Finds the first viable region for the proposed rect given the exclusion areas based upon x positions
+// Assumes writing direction is left to right, exclusionLines is sorted and preprocessed for y positions, and all inputs are valid
+static CGRect __FirstPossibleRectForProposed(CGRect proposed,
+                                             const std::vector<CGRect>& exclusionLines,
+                                             CGFloat padding,
+                                             CGFloat maxWidth) {
+    for (auto path : exclusionLines) {
+        if (proposed.origin.x < path.origin.x) {
+            proposed.size.width = path.origin.x - proposed.origin.x - padding;
+            if (proposed.size.width > padding) {
+                break;
+            }
+        }
+        if (path.origin.x + path.size.width > proposed.origin.x) {
+            proposed.origin.x = (path.origin.x + path.size.width) + padding;
+            proposed.size.width = maxWidth - proposed.origin.x;
+        }
+    }
+
+    return proposed;
+}
+
+// Finds the first point at which the given rect intersects the given exclusionContext, moving by delta each iteration
```

Comment:
I mean like, even if you're not wanting to do line intersections or whatever (would probably be a lot faster though), you can, off the top of my head:
- not calculate rect.origin.x + delta three times
- do a binary search instead of a linear search

how long do you expect this function to stick around? is it worth investing into a faster implementation at this/a later time?

---
In reply to: [90086723](https://github.com/Microsoft/WinObjC/pull/1398#discussion_r90086723) [](ancestors = 90086723)